OttDeBruin.pl, by Robin Meyers, 05mar2013
=========================================

<pre>
This program analyzes insertions and deletions in NGS amplicon sequence data.

Usage: bin/OttDeBruin.pl --metafile FILE (--in FILE | --indir DIR) --outdir DIR
		[--bcmismatch N] [--blastopt "-opt val"] [--threads N] [--ow]

Arguments (defaults in parentheses):

  --metafile       - Tab-delimited file containing experiment information

  --in             - Fasta file with multiplexed reads from sequencer

  --indir          - Directory containing demultiplexed fasta files

  --outdir         - Output directory

  --bcmismatch (1) - Number of mismatches allowed in matching barcode

  --blastopt (-task blastn-short -strand plus -gapopen 1 -gapextend 1 -reward 1 -penalty -4 -max_target_seqs 1) - Specify Blast options - refer to 'blastn -help'

  --threads (2)    - Number of processing core threads to run on

  --ow             - Overwrite output files if they already exist

  --help           - This helpful help screen.
</pre>


Meta file format
----------------
<pre>
The meta file must be a simple tab-delimited text file with the following headers:

experiment mid reference bindstart bindend cutstart cutend

Here is the text from a sample meta file ODB_meta.txt:

experiment         mid         reference  bindstart  bindend  cutstart  cutend
TALEN2_SCID058-IT  AGCACTGTAG  TALEN2.fa  74         123      92        107
TALEN2_neg_ctrl    ATCAGACACG  TALEN2.fa  74         123      92        107
TALEN3_SCID058-IT  TCGTCGCTCG  TALEN3.fa  206        255      223       238
TALEN3_neg_ctrl    ACATACGCGT  TALEN3.fa  206        255      223       238
TALEN4_SCID058-IT  CATAGTAGTG  TALEN4.fa  237        287      255       270
TALEN4_neg_ctrl    ATACGACGTA  TALEN4.fa  237        287      255       270
</pre>

Program Description
-------------------
<pre>
The multiplexed sequence file will be split up by MID, into experiment.fa files.
If --indir is given instead of --in, the program will look for experiment.fa files
in this directory, raising an error if any are not found.
The program will then look for the reference file (e.g. TALEN2.fa)
in the same folder as the sequence file(s).

The program then uses the blastn binary from the NCBI Blast+ toolkit to align
each sequence to the reference sequence.

Finally, the program parses this output, printing any insertion and deletion info
between the cut start and cut end coordinates of the reference sequence,
as well as the aligned sequences between the bind start and bind end coordinates.
This data goes into the experiment.txt file in the --outdir folder.

For example for the TALEN2_SCID058-IT experiment above, the output file will contain
aligned sequences between base 74 (inclusive) and 123 (exclusive), but will only count
indels that intersect the region between base 92 (inclusive) and 107 (exclusive).

The program prints out a summary file as well. The logic behind classifying each read
should be updated as desired in the parse_blast_output subroutine in lib/OttDeBruinHelper.pl
</pre>

Example Commands
----------------
<pre>

1)
OttDeBruin.pl --metafile ./in/ODB_meta.txt --in ./in/1.TCA.454ReadsLisa1_14_12.fna --outdir ./out/

Given this file tree
./ (working directory)
|-in/
|---ODB_meta.txt
|---1.TCA.454ReadsLisa1_14_12.fna
|---TALEN2.fa
|---TALEN3.fa
|---TALEN4.fa

This command will separate the input file into the six experiments above,
and run the default analysis.

This will be the resulting file tree
./ (working directory)
|-in/
|---ODB_meta.txt
|---1.TCA.454ReadsLisa1_14_12.fna
|---TALEN2.fa
|---TALEN3.fa
|---TALEN4.fa
|---TALEN2_SCID058-IT.fa
|---TALEN2_neg_ctrl.fa
|---TALEN3_SCID058-IT.fa
|---TALEN3_neg_ctrl.fa
|---TALEN4_SCID058-IT.fa
|---TALEN4_neg_ctrl.fa
|-out/
|---SummaryStats.txt
|---TALEN2_SCID058-IT.txt
|---TALEN2_neg_ctrl.txt
|---TALEN3_SCID058-IT.txt
|---TALEN3_neg_ctrl.txt
|---TALEN4_SCID058-IT.txt
|---TALEN4_neg_ctrl.txt
|---blasts/
|-----TALEN2_SCID058-IT_blast.txt
|-----TALEN2_neg_ctrl_blast.txt
|-----TALEN3_SCID058-IT_blast.txt
|-----TALEN3_neg_ctrl_blast.txt
|-----TALEN4_SCID058-IT_blast.txt
|-----TALEN4_neg_ctrl_blast.txt

2)
OttDeBruin.pl --metafile ./in/ODB_meta.txt --indir ./in/ --outdir ./out/
	--blastopt "-penalty -3" --threads 4 --ow

Given this file tree
./ (working directory)
|-in/
|---ODB_meta.txt
|---1.TCA.454ReadsLisa1_14_12.fna
|---TALEN2.fa
|---TALEN3.fa
|---TALEN4.fa
|---TALEN2_SCID058-IT.fa
|---TALEN2_neg_ctrl.fa
|---TALEN3_SCID058-IT.fa
|---TALEN3_neg_ctrl.fa
|---TALEN4_SCID058-IT.fa
|---TALEN4_neg_ctrl.fa

This command skips the demultiplexing steps and goes straight to analyzing
the reads. In contrast to the first example, this commmand will overwrite
any existing output files, changes the blast options to use a mismatch penalty
of -3, and runs on 4 threads instead of the default 2. The resulting file tree
will be the same as in example 1.
</pre>
